src/app/clusters/electrical-energy-measurement-server/CodegenIntegration.cpp

```diff
@@ -0,0 +1,216 @@
+    VerifyOrReturnValue(cluster->Features().Has(Feature::kPeriodicEnergy), false);
+
+    Events::PeriodicEnergyMeasured::Type event;
+
+    if (cluster->Features().Has(Feature::kImportedEnergy))
+    {
+        cluster->SetPeriodicEnergyImported(energyImported);
+        event.energyImported = energyImported;
+    }
+
+    if (cluster->Features().Has(Feature::kExportedEnergy))
+    {
+        cluster->SetPeriodicEnergyExported(energyExported);
+        event.energyExported = energyExported;
+    }
+
+    EventNumber eventNumber;
+    CHIP_ERROR error = app::LogEvent(event, endpointId, eventNumber);
+    if (CHIP_NO_ERROR != error)
+    {
+#if defined(CHIP_CODEGEN_CONFIG_ENABLE_CODEGEN_INTEGRATION_LOOKUP_ERRORS) &&                                                       \
+    CHIP_CODEGEN_CONFIG_ENABLE_CODEGEN_INTEGRATION_LOOKUP_ERRORS
+        ChipLogError(Zcl, "[NotifyPeriodicEnergyMeasured] Unable to send event: %" CHIP_ERROR_FORMAT " [endpointId=%d]",
+                     error.Format(), endpointId);
+#endif // CHIP_CODEGEN_CONFIG_ENABLE_CODEGEN_INTEGRATION_LOOKUP_ERRORS
+        return false;
+    }
+    ChipLogProgress(Zcl, "[NotifyPeriodicEnergyMeasured] Sent event [endpointId=%d,eventNumber=%lu]", endpointId,
+                    static_cast<unsigned long>(eventNumber));
+    return true;
+}
+
+CHIP_ERROR ElectricalEnergyMeasurementAttrAccess::Init()
+{
+    CHIP_ERROR err = CodegenDataModelProvider::Instance().Registry().Register(mCluster.Registration());
+
+    if (err != CHIP_NO_ERROR)
+    {
+#if defined(CHIP_CODEGEN_CONFIG_ENABLE_CODEGEN_INTEGRATION_LOOKUP_ERRORS) &&                                                       \
+    CHIP_CODEGEN_CONFIG_ENABLE_CODEGEN_INTEGRATION_LOOKUP_ERRORS
+        ChipLogError(AppServer, "Failed to register cluster %u/" ChipLogFormatMEI ": %" CHIP_ERROR_FORMAT, mEndpointId,
+                     ChipLogValueMEI(ElectricalEnergyMeasurement::Id), err.Format());
+#endif // CHIP_CODEGEN_CONFIG_ENABLE_CODEGEN_INTEGRATION_LOOKUP_ERRORS
+    }
+    return err;
+}
+
+void ElectricalEnergyMeasurementAttrAccess::Shutdown()
+{
+    CHIP_ERROR err = CodegenDataModelProvider::Instance().Registry().Unregister(&(mCluster.Cluster()));
+
+    if (err != CHIP_NO_ERROR)
+    {
+#if defined(CHIP_CODEGEN_CONFIG_ENABLE_CODEGEN_INTEGRATION_LOOKUP_ERRORS) &&                                                       \
+    CHIP_CODEGEN_CONFIG_ENABLE_CODEGEN_INTEGRATION_LOOKUP_ERRORS
+        ChipLogError(AppServer, "Failed to unregister cluster %u/" ChipLogFormatMEI ": %" CHIP_ERROR_FORMAT, mEndpointId,
+                     ChipLogValueMEI(ElectricalEnergyMeasurement::Id), err.Format());
+#endif // CHIP_CODEGEN_CONFIG_ENABLE_CODEGEN_INTEGRATION_LOOKUP_ERRORS
+    }
+}
+
+bool ElectricalEnergyMeasurementAttrAccess::HasFeature(Feature aFeature) const
+{
+    return const_cast<RegisteredServerCluster<ElectricalEnergyMeasurementCluster> &>(mCluster).Cluster().Features().Has(aFeature);
```

Comment:
Why do we need a const_cast here?  This suggests that something has non-const-correct APIs; we should not need to have a non-const thing to do a feature map check.